### PR TITLE
Able Minecart to move. #WIP

### PR DESCRIPTION
**Pull request by NycuRO**
This PR isn't ready. Please wait for translate. Codes it's from Genisys. Add more.

If exist errors ,fix . If i don't know ,i want help.

Anyone can write code and correct me :) 

Thanks and please not close PR. :)

### CODE REVIEW

File: src/main/java/cn/nukkit/entity/item/EntityMinecartEmpty.java [hunk 1]
```diff
@@ -340,85 +340,141 @@ private void checkForVertical(railType, currentDirection) {
...
+		rails = [];
+		for (z = minZ; z <= maxZ; ++z) {
+			for (x = minX; x <= maxX; ++x) {
+				for (y = minY; y <= maxY; ++y) {
+					Block block = this.level.getBlock(this.temporalVector.setComponents(x, y, z));
+					if (block.getId(), {Block::RAIL, Block::ACTIVATOR_RAIL, Block::DETECTOR_RAIL, Block::POWERED_RAIL}) rails[] = block;
+				}
+			}
+		}
+		minDistance = PHP_INT_MAX;
```
Comment (NycuRO) [inline src/main/java/cn/nukkit/entity/item/EntityMinecartEmpty.java]: Anyone can help me this? What code uses " PHP_INT_MAX " in java?

Comment (IrishUHC) [inline src/main/java/cn/nukkit/entity/item/EntityMinecartEmpty.java]: Integer.MAX_VALUE, you can cast if need be, I'm not sure whether it returns a float or not.

**Comment by IrishUHC** [comment]
Change the name to "Allow minecarts to move", instead of its current name?

**Comment by NycuRO** [comment]
@IrishUHC I don't understand what you speak. You can add to line code?

**Comment by NycuRO** [comment]
Anyone can help me to more translate? Thanks so much :)

**Comment by liach** [comment]
You should not open a pr if you do not have an ide, or if you do not know about java.

**Comment by NycuRO** [comment]
@liach i want to create nukkit more more used,but i need help because i'm starter on java. If anyone can help me ,can :) If no,no. Devs can translate from *other software* ,but not translate. I'm translate 50-60% ,and other 50-40% ,can devs for more easy translation :) 👍

File: src/main/java/cn/nukkit/entity/item/EntityMinecartEmpty.java [hunk 2]
```diff
@@ -15,12 +25,25 @@
...
     public static final int DATA_VEHICLE_DISPLAY_DATA = 20;
     public static final int DATA_VEHICLE_DISPLAY_OFFSET = 21;
     public static final int DATA_VEHICLE_CUSTOM_DISPLAY = 22;
-
-    // TODO: 2016/1/30 check if these numbers correct
+    
+    private String int state = Minecart.STATE_INITIAL;
+    private String int direction = -1;
+    private String int moveVector = [];
+    private String int requestedPosition = null;
```
Comment (liach) [inline src/main/java/cn/nukkit/entity/item/EntityMinecartEmpty.java]: compile errors!

File: src/main/java/cn/nukkit/entity/item/EntityMinecartEmpty.java [hunk 3]
```diff
@@ -45,24 +68,418 @@ protected float getGravity() {
...
+	moveVector[Entity.EAST] = new Vector3(0, 0, -1);
+	moveVector[Entity.WEST] = new Vector3(0, 0, 1);
         super.initEntity();
     }
 
     @Override
     public boolean onUpdate(int currentTick) {
-        // TODO: 2016/1/30 run run run!
-        // TODO: 2016/1/30 split to onXXXRailPass, such as, protected void onActivatorRailPass(...)
+        if (this.closed != false) {
```
Comment (liach) [inline src/main/java/cn/nukkit/entity/item/EntityMinecartEmpty.java]:
```java
if (this.closed)
```

File: src/main/java/cn/nukkit/entity/item/EntityMinecartEmpty.java [hunk 4]
```diff
@@ -45,24 +68,418 @@ protected float getGravity() {
...
+		for (y = -1; y != 2 && this.state == Minecart.STATE_INITIAL; y++) {
+			positionToCheck = this.temporalVector.setComponents(this.x, this.y + y, this.z);
+			Block block = this.level.getBlock(positionToCheck);
+			if (this.isRail(block)) {
+				minecartPosition = positionToCheck.floor().add(0.5, 0, 0.5);
+				this.setPosition(minecartPosition);    // Move minecart to center of rail
+				this.state = Minecart::STATE_ON_RAIL;
+			}
+		}
+		if (this.state != Minecart::STATE_ON_RAIL) {
```
Comment (liach) [inline src/main/java/cn/nukkit/entity/item/EntityMinecartEmpty.java]:
```java
Minecart.STATE_ON_RAIL
```
This is not C or C++

File: src/main/java/cn/nukkit/entity/item/EntityMinecartEmpty.java [hunk 5]
```diff
@@ -45,24 +68,418 @@ protected float getGravity() {
...
+			this.state = Minecart::STATE_OFF_RAIL;
+		}
+	}
+    
+    @Override
+    private void isRail(rail) {
```
Comment (liach) [inline src/main/java/cn/nukkit/entity/item/EntityMinecartEmpty.java]: private methods can never be overridden.

Comment (liach) [inline src/main/java/cn/nukkit/entity/item/EntityMinecartEmpty.java]: And what is `rail`?

File: src/main/java/cn/nukkit/entity/item/EntityMinecartEmpty.java [hunk 6]
```diff
@@ -45,24 +68,418 @@ protected float getGravity() {
...
+		minZ = NukkitMath.floorFloat(this.boundingBox.minZ);
+		maxX = NukkitMath.ceilFloat(this.boundingBox.maxX);
+		maxY = NukkitMath.ceilFloat(this.boundingBox.maxY);
+		maxZ = NukkitMath.ceilFloat(this.boundingBox.maxZ);
+		rails = [];
+		for (z = minZ; z <= maxZ; ++z) {
+			for (x = minX; x <= maxX; ++x) {
+				for (y = minY; y <= maxY; ++y) {
+					Block block = this.level.getBlock(this.temporalVector.setComponents(x, y, z));
+					if (block.getId(), {Block.RAIL, Block.ACTIVATOR_RAIL, Block.DETECTOR_RAIL, Block.POWERED_RAIL}) rails[] = block;
```
Comment (liach) [inline src/main/java/cn/nukkit/entity/item/EntityMinecartEmpty.java]: ?!

**Comment by liach** [comment]
@NycuRO I sincerely suggest you to stop modifying code on GitHub. At least you should learn to use GitHub for Windows/Mac first.

File: src/main/java/cn/nukkit/entity/item/EntityMinecartEmpty.java (hunk 2 above)
Comment (NycuRO) [inline src/main/java/cn/nukkit/entity/item/EntityMinecartEmpty.java]: I know. I don;t know to fix here. Anyone can help me?

**Comment by Snake1999** [comment]
Are you serious? Java programs can not be written uses copy and paste :/

**Comment by NycuRO** [comment]
@Snake1999 If not all implements this..

File: src/main/java/cn/nukkit/entity/item/EntityMinecartEmpty.java (hunk 6 above)
Comment (NycuRO) [inline src/main/java/cn/nukkit/entity/item/EntityMinecartEmpty.java]: @liach 
	private function moveIfRail(){
		$nextMoveVector = $this->moveVector[$this->direction];
		$nextMoveVector = $nextMoveVector->multiply($this->moveSpeed);
		$newVector = $this->add($nextMoveVector->x, $nextMoveVector->y, $nextMoveVector->z);
		$possibleRail = $this->getCurrentRail();
		if(in_array($possibleRail->getId(), [Block::RAIL, Block::ACTIVATOR_RAIL, Block::DETECTOR_RAIL, Block::POWERED_RAIL])) {
			$this->moveUsingVector($newVector);
			return true;
		}
	}

**Comment by NycuRO** [comment]
@Snake1999 If you like help me to translate and fix errors, 
I would be grateful. And other can :)

**Comment by Snake1999** [comment]
Translate?/

**Comment by Creeperface01** [comment]
@NycuRO test your code before you create pull request

**Comment by NycuRO** [comment]
@Snake1999 Correct errors*

@Creeperface01 Is created because help me anyone to fix errors and implement Minecart...

**Comment by Creeperface01** [comment]
IDE shows you all syntax errors, so you can fix them by yourself

File: src/main/java/cn/nukkit/entity/item/EntityMinecartEmpty.java (hunk 4 above)
Comment (Creeperface01) [inline src/main/java/cn/nukkit/entity/item/EntityMinecartEmpty.java]: PHP*

**Comment by NycuRO** [comment]
@Creeperface01 I don't know how fix all..

**Comment by Creeperface01** [comment]
Search on Google for right syntax

**Comment by liach** [comment]
@Snake1999 这个pr太呵呵了，还是关了吧

**Comment by NycuRO** [comment]
Anyone can help me ? I don't know how fix.... Developers? Or close ?... @Creeperface01 @MrPowerGamerBR

File: src/main/java/cn/nukkit/entity/item/EntityMinecartEmpty.java (hunk 6 above)
Comment (TheDiamondYT1) [inline src/main/java/cn/nukkit/entity/item/EntityMinecartEmpty.java]: Php nub

**Comment by TheDiamondYT1** [comment]
Everything about this is wrong. Please close.

**Comment by liach** [comment]
:trollface: Meh this really need to be closed. You should not show how bad you are in Java with such pull requests.

**Comment by TheDiamondYT1** [comment]
@Liach He made som PR's literally for him to test stuff

File: src/main/java/cn/nukkit/entity/item/EntityMinecartEmpty.java (hunk 6 above)
Comment (liach) [inline src/main/java/cn/nukkit/entity/item/EntityMinecartEmpty.java]: Well I am not a web programmer. You should not code java in the way you code a web language.

**Comment by liach** [comment]
@NycuRO Take a look at this before opening another pull request: http://www.freejavaguide.com/corejava.htm

**Comment by TheDiamondYT1** [comment]
:+1:

**Pull request closed without merging**